File: infiniteSearch.java
//Search in Infinite sorted array
//Time Complexity: O(log n)
//Space Complexity: O(1)

public class infiniteSearch {
    public static void main(String[] args) {
        int[] a = {3,5,6,7,8};
        int key = 7;

        int index = modBinSearch(a, key);
        System.out.println(index);
    }


    private static int binarySearch(int a[], int key, int low, int high) {
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (a[mid] == key)
                return mid;
            else if (a[mid] > key)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return -1;
    }

    private static int modifyBinarySearch(int a[], int key){
        int low = 0;
        int high = 1;
        int pHolder = a[0];

        while(pHolder < key){
            low = high;
            high = high * 2;
            pHolder = a[high];
        }

        return binarySearch(a, key, low, high);
    }

}

File: Infinitesearch.py
# // Time Complexity : O(log n)
# // Space Complexity : O(1) as no extra space being used
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach
'''
Here we follow increasing the search space twice approach
first we need to find the range in which the target present
  we took two pointers low and high, move high in steps of 2 times prev position and low to previous high position
Then apply binary search on that range

'''

# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        low = 0
        high = 1
        while reader.get(high) < target:
            low = high
            high *= 2
        while low <= high:
            mid = low + (high-low)//2
            if reader.get(mid) == target:
                return mid
            elif reader.get(mid) <= target:
                low = mid + 1
            else:
                high = mid - 1
        return -1
            
        
